### PR TITLE
[13.x] Support OAuth2 Server v9

### PR DESCRIPTION
**Pull request by hafezdivandari**
Support [OAuth2 Server v9](https://github.com/thephpleague/oauth2-server/releases/tag/9.0.0).

- [x] thephpleague/oauth2-server#1398
- [x] thephpleague/oauth2-server#1402

### CODE REVIEW

**Comment by hafezdivandari** [comment]
@driesvints this repo doesn't have `master` branch?

**Comment by jayan-blutui** [comment]
Great work on creating this PR so quickly seeing v9rc just launched a couple of days ago.

I know this PR is still a WIP but do you plan to include the new Device Authorization?

**Comment by hafezdivandari** [comment]
@jayan-blutui thanks.

> I know this PR is still a WIP but do you plan to include the new Device Authorization?

Not in this PR, as it makes this hard to review.

**Comment by driesvints** [comment]
Thanks a lot for this PR @hafezdivandari. We should indeed try to support the new Device flow in v13 but let's indeed tackle that in a subsequent PR.

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,6 @@
 composer.lock
 /phpunit.xml
 .phpunit.result.cache
+/.idea
+/.vscode
+/.rnd
```
Comment (driesvints) [inline .gitignore]: I'd prefer if you removed these and added these to a global gitignore file: https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files

File: composer.json [hunk 2]
```diff
@@ -26,18 +27,18 @@
         "illuminate/encryption": "^9.21|^10.0|^11.0",
         "illuminate/http": "^9.21|^10.0|^11.0",
         "illuminate/support": "^9.21|^10.0|^11.0",
```
Comment (driesvints) [inline composer.json:29]: In a next PR we may drop Laravel v9 and v10 support as well as PHP 8.1

File: src/Bridge/FormatsScopesForStorage.php [hunk 3]
```diff
@@ -2,26 +2,27 @@
...
     {
         return json_encode($this->scopesToArray($scopes));
     }
 
     /**
      * Get an array of scope identifiers for storage.
      *
-     * @param  array  $scopes
-     * @return array
+     * @param  ScopeEntityInterface[]  $scopes
```
Comment (driesvints) [inline src/Bridge/FormatsScopesForStorage.php]: Can we use FQN here?

**Review by driesvints — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -16,13 +16,9 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        php: ['8.0', 8.1, 8.2, 8.3]
+        php: [8.1, 8.2, 8.3]
         laravel: [9, 10, 11]
```
Comment (driesvints) [inline .github/workflows/tests.yml:20]: In a next PR we may drop Laravel v9 and v10 support as well as PHP 8.1

File: composer.json [hunk 5]
```diff
@@ -26,18 +27,18 @@
...
-        "lcobucci/jwt": "^4.3|^5.0",
-        "league/oauth2-server": "^8.5.3",
+        "lcobucci/jwt": "^5.0",
+        "league/oauth2-server": "dev-fix-passport-compatibility",
```
Comment (hafezdivandari) [inline composer.json]: TBD

**Comment by hafezdivandari** [comment]
@driesvints This PR is ready, but it should target `master`, which doesn't exist on this repo.

**Comment by driesvints** [comment]
@hafezdivandari that's okay. We're using a different approach these days. I'll talk to the team to prepare a new major version. Ideally we'd also take in the new device flow (separate PR).

Thanks for working on this one!

**Review by driesvints — APPROVED** [review]

**Comment by driesvints** [comment]
@hafezdivandari I retargeted your PR to 13.x now

**Comment by siarheipashkevich** [comment]
@hafezdivandari @driesvints why not to 12.x?

**Comment by driesvints** [comment]
@siarheipashkevich there's just too many breaking changes in here.

@hafezdivandari reminds me: could you make a list of upgrade steps in `UPGRADE.md`? I think we only need to reference the OAuth Server client v9 upgrade and a link to its own release notes, telling people to review those as well: https://github.com/thephpleague/oauth2-server/blob/master/CHANGELOG.md#900---released-2024-05-13

**Comment by taylorotwell** [comment]
Can someone give me an idea of the full upgrade steps and mark as ready for review?

**Comment by driesvints** [comment]
@taylorotwell those are already added to the upgrade guide in this PR. The main reason for this new major version is the oauth server v9 bump and the method signature changes.

**Comment by hafezdivandari** [comment]
@driesvints We are not going to release 13.x immediately after merging this right? I have 1 or 2 PRs for 13.x before release. I don't want to add too many changes at once, that makes this one hard to review.

**Comment by driesvints** [comment]
@hafezdivandari no I'll wait for those

**Comment by taylorotwell** [comment]
@driesvints I guess I'm looking for a general summary of the changes someone will need to make. What changes will a typical Laravel Passport app need to make? Earlier it was mentioned there are too many breaking changes for patch - what are those breaking changes specifically?

**Comment by hafezdivandari** [comment]
@taylorotwell nothing, just PHP 8.1+ and removed `message` property from `OAuthException` HTTP response. Now just use `error_description` as per the OAuth 2 spec.

The braking change is that most of the methods signatures are changed.

**Pull request merged**